{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "bc0114b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from lxml import html\n",
    "import requests\n",
    "import json\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import MACCSkeys\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "7770f205",
   "metadata": {},
   "outputs": [],
   "source": [
    "class KEGGCompound:\n",
    "    def __init__(self, id):\n",
    "        self.id = id\n",
    "        self.__molFileText = None\n",
    "        self.__molFileFetched = False\n",
    "        #self.mol = None\n",
    "        self.pathways = []\n",
    " \n",
    "    #\n",
    "    # Get the SMILES string associated with this compound.\n",
    "    #\n",
    "    def getSMILES(self):\n",
    "        # Use RDKit to convert the mol file string to a SMILES. \n",
    "        mol = Chem.MolFromMolBlock(self.fetchMOLFileString())\n",
    "        try:\n",
    "            smilesString = Chem.rdmolfiles.MolToSmiles(mol)\n",
    "        except:\n",
    "            return(\"Unhandled\")        # This was a polymer, which RDKit can't handle.\n",
    "        return(smilesString)\n",
    "\n",
    "    #\n",
    "    # Get the MACCS fingerprint associated with this compound.\n",
    "    #\n",
    "    def getMACCS(self):\n",
    "        mol = Chem.MolFromMolBlock(self.fetchMOLFileString())\n",
    "        maccsFp = MACCSkeys.FingerprintMol(mol).ToBitString()\n",
    "        return(maccsFp)\n",
    "\n",
    "    #\n",
    "    #\n",
    "    #\n",
    "    def fetchMOLFileString(self):\n",
    "        if (self.__molFileFetched == True):\n",
    "            return(self.__molFileText)\n",
    "        urlStr = \"https://www.genome.jp/dbget-bin/www_bget?-f+m+compound+\" + self.id\n",
    "        molFile = requests.get(urlStr)\n",
    "        self.__molFileText = str(molFile.text)\n",
    "        self.__molFileFetched = True\n",
    "        #self.mol = Chem.MolFromMolBlock(self.__molFileText)\n",
    "        return(self.__molFileText)\n",
    "\n",
    "    #\n",
    "    #\n",
    "    #\n",
    "    def addPathway(self, mod):\n",
    "        self.pathways.append(mod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ca856fe1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "class KEGGPathway:\n",
    "    def __init__(self, id, name, className):\n",
    "        self.id = id\n",
    "        self.name = name\n",
    "        self.className = className\n",
    "        self.compounds = []\n",
    "        self.__compoundsFetched = False\n",
    "\n",
    "    #\n",
    "    # REST API URL to get compounds linked to this pathway\n",
    "    #\n",
    "    def getCompoundsURL(self):\n",
    "        return f\"https://rest.kegg.jp/link/compound/path:map{self.id}\"\n",
    "\n",
    "    #\n",
    "    # Fetch a list of KEGGCompound objects associated with this pathway\n",
    "    #\n",
    "    def fetchCompounds(self):\n",
    "        if self.__compoundsFetched:\n",
    "            return self.compounds\n",
    "\n",
    "        retval = []\n",
    "        url = self.getCompoundsURL()\n",
    "        response = requests.get(url)\n",
    "        response.raise_for_status()  # raise error if request fails\n",
    "\n",
    "        # Each line looks like: \"path:map00010\\tcpd:C00022\"\n",
    "        lines = response.text.strip().splitlines()\n",
    "        compoundIDs = [line.split(\"\\t\")[1].split(\":\")[1] for line in lines]\n",
    "\n",
    "        for cid in compoundIDs:\n",
    "            retval.append(KEGGCompound(cid))  # assuming KEGGCompound class exists\n",
    "\n",
    "        self.compounds = retval\n",
    "        self.__compoundsFetched = True\n",
    "        return retval\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e411a9f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getPathways():\n",
    "    retval = []        # This will be a list of KEGGModule objects.\n",
    "    # Get the KEGG modules page as a string.\n",
    "    #page = requests.get(\"https://www.genome.jp/kegg-bin/download_htext?htext=ko00002.keg&format=json&filedir=\")\n",
    "    #jsonFile = open(\"../../data/kegg-data/ko00002.json\")\n",
    "    #jsonStr = jsonFile.read()\n",
    "    #jsonFile.close()\n",
    "    #print(page.text)\n",
    "    #jsonData = json.loads(jsonStr)\n",
    "    #print(jsonData.keys())\n",
    "    #print(type(jsonData))\n",
    "    #tree = html.fromstring(page.content)\n",
    "\n",
    "    page = requests.get(\"https://www.genome.jp/kegg/pathway.html\")\n",
    "    tree = html.fromstring(page.content)\n",
    "    #.xpath('//div[@id=\"definition\"]/table/tr/td[text()=\"Reaction\"]/../td[2]/a')\n",
    "    for bNode in tree.xpath('//b'):\n",
    "        pathwayClassNum = bNode.xpath('text()')[0].split(' ')[0]\n",
    "        pathwayClassName = bNode.xpath('text()')[0].split(' ')[1]\n",
    "\n",
    "        if (pathwayClassNum != '1.0' and pathwayClassNum != '1.' and pathwayClassNum != '1.12' and pathwayClassNum[0] == '1'):\n",
    "            print(\"b node = \" + bNode.xpath('text()')[0])\n",
    "            print(\"Pathway class number and name: \" + pathwayClassNum + \", \" + pathwayClassName)\n",
    "\n",
    "            # We are interested in this pathway.\n",
    "            pathwayIDNodes = bNode.xpath('following-sibling::div')[0].xpath('dl/dt')\n",
    "            for pathwayIDNode in pathwayIDNodes:\n",
    "                # Get the id.\n",
    "                pathwayID = pathwayIDNode.xpath('text()')[0]\n",
    "                pathwayName = pathwayIDNode.xpath('following-sibling::dd/a/text()')[0]\n",
    "                print(\"Pathway id and name: \" + str(pathwayID) + \", \" + str(pathwayName))\n",
    "                newPathway = KEGGPathway(pathwayID, pathwayName, pathwayClassName)\n",
    "                retval.append(newPathway)\n",
    "        #print(tree.xpath('//b/text()'))\n",
    "    return(retval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbfde17b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b node = 1.1 Carbohydrate metabolism\n",
      "Pathway class number and name: 1.1, Carbohydrate\n",
      "Pathway id and name: 00010 , Glycolysis / Gluconeogenesis\n",
      "Pathway id and name: 00020 , Citrate cycle (TCA cycle)\n",
      "Pathway id and name: 00030 , Pentose phosphate pathway\n",
      "Pathway id and name: 00040 , Pentose and glucuronate interconversions\n",
      "Pathway id and name: 00051 , Fructose and mannose metabolism\n",
      "Pathway id and name: 00052 , Galactose metabolism\n",
      "Pathway id and name: 00053 , Ascorbate and aldarate metabolism\n",
      "Pathway id and name: 00500 , Starch and sucrose metabolism\n",
      "Pathway id and name: 00620 , Pyruvate metabolism\n",
      "Pathway id and name: 00630 , Glyoxylate and dicarboxylate metabolism\n",
      "Pathway id and name: 00640 , Propanoate metabolism\n",
      "Pathway id and name: 00650 , Butanoate metabolism\n",
      "Pathway id and name: 00660 , C5-Branched dibasic acid metabolism\n",
      "Pathway id and name: 00562 , Inositol phosphate metabolism\n",
      "b node = 1.2 Energy metabolism\n",
      "Pathway class number and name: 1.2, Energy\n",
      "Pathway id and name: 00190 , Oxidative phosphorylation\n",
      "Pathway id and name: 00195 , Photosynthesis\n",
      "Pathway id and name: 00196, Photosynthesis - antenna proteins\n",
      "Pathway id and name: 00710 , Carbon fixation by Calvin cycle\n",
      "Pathway id and name: 00720 , Other carbon fixation pathways\n",
      "Pathway id and name: 00680 , Methane metabolism\n",
      "Pathway id and name: 00910 , Nitrogen metabolism\n",
      "Pathway id and name: 00920 , Sulfur metabolism\n",
      "b node = 1.3 Lipid metabolism\n",
      "Pathway class number and name: 1.3, Lipid\n",
      "Pathway id and name: 00061 , Fatty acid biosynthesis\n",
      "Pathway id and name: 00062 , Fatty acid elongation\n",
      "Pathway id and name: 00071 , Fatty acid degradation\n",
      "Pathway id and name: 00073, Cutin, suberine and wax biosynthesis\n",
      "Pathway id and name: 00074 , Mycolic acid biosynthesis\n",
      "Pathway id and name: 00100 , Steroid biosynthesis\n",
      "Pathway id and name: 00120 , Primary bile acid biosynthesis\n",
      "Pathway id and name: 00121, Secondary bile acid biosynthesis\n",
      "Pathway id and name: 00140 , Steroid hormone biosynthesis\n",
      "Pathway id and name: 00561 , Glycerolipid metabolism\n",
      "Pathway id and name: 00564 , Glycerophospholipid metabolism\n",
      "Pathway id and name: 00565 , Ether lipid metabolism\n",
      "Pathway id and name: 00600 , Sphingolipid metabolism\n",
      "Pathway id and name: 00590, Arachidonic acid metabolism\n",
      "Pathway id and name: 00591, Linoleic acid metabolism\n",
      "Pathway id and name: 00592 , alpha-Linolenic acid metabolism\n",
      "Pathway id and name: 01040 , Biosynthesis of unsaturated fatty acids\n",
      "b node = 1.4 Nucleotide metabolism\n",
      "Pathway class number and name: 1.4, Nucleotide\n",
      "Pathway id and name: 00230 , Purine metabolism\n",
      "Pathway id and name: 00240 , Pyrimidine metabolism\n",
      "b node = 1.5 Amino acid metabolism\n",
      "Pathway class number and name: 1.5, Amino\n",
      "Pathway id and name: 00250 , Alanine, aspartate and glutamate metabolism\n",
      "Pathway id and name: 00260 , Glycine, serine and threonine metabolism\n",
      "Pathway id and name: 00270 , Cysteine and methionine metabolism\n",
      "Pathway id and name: 00280 , Valine, leucine and isoleucine degradation\n",
      "Pathway id and name: 00290 , Valine, leucine and isoleucine biosynthesis\n",
      "Pathway id and name: 00300 , Lysine biosynthesis\n",
      "Pathway id and name: 00310 , Lysine degradation\n",
      "Pathway id and name: 00220 , Arginine biosynthesis\n",
      "Pathway id and name: 00330 , Arginine and proline metabolism\n",
      "Pathway id and name: 00340 , Histidine metabolism\n",
      "Pathway id and name: 00350 , Tyrosine metabolism\n",
      "Pathway id and name: 00360 , Phenylalanine metabolism\n",
      "Pathway id and name: 00380 , Tryptophan metabolism\n",
      "Pathway id and name: 00400 , Phenylalanine, tyrosine and tryptophan biosynthesis\n",
      "b node = 1.6 Metabolism of other amino acids\n",
      "Pathway class number and name: 1.6, Metabolism\n",
      "Pathway id and name: 00410 , beta-Alanine metabolism\n",
      "Pathway id and name: 00430 , Taurine and hypotaurine metabolism\n",
      "Pathway id and name: 00440, Phosphonate and phosphinate metabolism\n",
      "Pathway id and name: 00450, Selenocompound metabolism\n",
      "Pathway id and name: 00460 , Cyanoamino acid metabolism\n",
      "Pathway id and name: 00470 , D-Amino acid metabolism\n",
      "Pathway id and name: 00480 , Glutathione metabolism\n",
      "b node = 1.7 Glycan biosynthesis and metabolism\n",
      "Pathway class number and name: 1.7, Glycan\n",
      "Pathway id and name: 00520 , Amino sugar and nucleotide sugar metabolism\n",
      "Pathway id and name: 00541 , Biosynthesis of various nucleotide sugars\n",
      "Pathway id and name: 00510 , N-Glycan biosynthesis\n",
      "Pathway id and name: 00513 , Various types of N-glycan biosynthesis\n",
      "Pathway id and name: 00512 , Mucin type O-glycan biosynthesis\n",
      "Pathway id and name: 00515 , Mannose type O-glycan biosynthesis\n",
      "Pathway id and name: 00514, Other types of O-glycan biosynthesis\n",
      "Pathway id and name: 00532 , Glycosaminoglycan biosynthesis - chondroitin sulfate / dermatan sulfate\n",
      "Pathway id and name: 00534 , Glycosaminoglycan biosynthesis - heparan sulfate / heparin\n",
      "Pathway id and name: 00533, Glycosaminoglycan biosynthesis - keratan sulfate\n",
      "Pathway id and name: 00531 , Glycosaminoglycan degradation\n",
      "Pathway id and name: 00563 , Glycosylphosphatidylinositol (GPI)-anchor biosynthesis\n",
      "Pathway id and name: 00601 , Glycosphingolipid biosynthesis - lacto and neolacto series\n",
      "Pathway id and name: 00603 , Glycosphingolipid biosynthesis - globo and isoglobo series\n",
      "Pathway id and name: 00604 , Glycosphingolipid biosynthesis - ganglio series\n",
      "Pathway id and name: 00511, Other glycan degradation\n",
      "Pathway id and name: 00540 , Lipopolysaccharide biosynthesis\n",
      "Pathway id and name: 00542, O-Antigen repeat unit biosynthesis\n",
      "Pathway id and name: 00550, Peptidoglycan biosynthesis\n",
      "Pathway id and name: 00552, Teichoic acid biosynthesis\n",
      "Pathway id and name: 00571, Lipoarabinomannan (LAM) biosynthesis\n",
      "Pathway id and name: 00572 , Arabinogalactan biosynthesis - Mycobacterium\n",
      "Pathway id and name: 00543, Exopolysaccharide biosynthesis\n",
      "b node = 1.8 Metabolism of cofactors and vitamins\n",
      "Pathway class number and name: 1.8, Metabolism\n",
      "Pathway id and name: 00730 , Thiamine metabolism\n",
      "Pathway id and name: 00740 , Riboflavin metabolism\n",
      "Pathway id and name: 00750 , Vitamin B6 metabolism\n",
      "Pathway id and name: 00760 , Nicotinate and nicotinamide metabolism\n",
      "Pathway id and name: 00770 , Pantothenate and CoA biosynthesis\n",
      "Pathway id and name: 00780 , Biotin metabolism\n",
      "Pathway id and name: 00785 , Lipoic acid metabolism\n",
      "Pathway id and name: 00790 , Folate biosynthesis\n",
      "Pathway id and name: 00670 , One carbon pool by folate\n",
      "Pathway id and name: 00830, Retinol metabolism\n",
      "Pathway id and name: 00860 , Porphyrin metabolism\n",
      "Pathway id and name: 00130 , Ubiquinone and other terpenoid-quinone biosynthesis\n",
      "b node = 1.9 Metabolism of terpenoids and polyketides\n",
      "Pathway class number and name: 1.9, Metabolism\n",
      "Pathway id and name: 00900 , Terpenoid backbone biosynthesis\n",
      "Pathway id and name: 00902, Monoterpenoid biosynthesis\n",
      "Pathway id and name: 00909, Sesquiterpenoid and triterpenoid biosynthesis\n",
      "Pathway id and name: 00904 , Diterpenoid biosynthesis\n",
      "Pathway id and name: 00906 , Carotenoid biosynthesis\n",
      "Pathway id and name: 00905 , Brassinosteroid biosynthesis\n",
      "Pathway id and name: 00981, Insect hormone biosynthesis\n",
      "Pathway id and name: 00908, Zeatin biosynthesis\n",
      "Pathway id and name: 00903, Limonene degradation\n",
      "Pathway id and name: 00907, Pinene, camphor and geraniol degradation\n",
      "Pathway id and name: 01052, Type I polyketide structures\n",
      "Pathway id and name: 00522 , Biosynthesis of 12-, 14- and 16-membered macrolides\n",
      "Pathway id and name: 01051, Biosynthesis of ansamycins\n",
      "Pathway id and name: 01059 , Biosynthesis of enediyne antibiotics\n",
      "Pathway id and name: 01056 , Biosynthesis of type II polyketide backbone\n",
      "Pathway id and name: 01057 , Biosynthesis of type II polyketide products\n",
      "Pathway id and name: 00253 , Tetracycline biosynthesis\n",
      "Pathway id and name: 00523 , Polyketide sugar unit biosynthesis\n",
      "Pathway id and name: 01054, Nonribosomal peptide structures\n",
      "Pathway id and name: 01053, Biosynthesis of siderophore group nonribosomal peptides\n",
      "Pathway id and name: 01055, Biosynthesis of vancomycin group antibiotics\n",
      "b node = 1.10 Biosynthesis of other secondary metabolites\n",
      "Pathway class number and name: 1.10, Biosynthesis\n",
      "Pathway id and name: 00940 , Phenylpropanoid biosynthesis\n",
      "Pathway id and name: 00945, Stilbenoid, diarylheptanoid and gingerol biosynthesis\n",
      "Pathway id and name: 00941 , Flavonoid biosynthesis\n",
      "Pathway id and name: 00944, Flavone and flavonol biosynthesis\n",
      "Pathway id and name: 00942, Anthocyanin biosynthesis\n",
      "Pathway id and name: 00943 , Isoflavonoid biosynthesis\n",
      "Pathway id and name: 00946 , Degradation of flavonoids\n",
      "Pathway id and name: 00901 , Indole alkaloid biosynthesis\n",
      "Pathway id and name: 00403 , Indole diterpene alkaloid biosynthesis\n",
      "Pathway id and name: 00950 , Isoquinoline alkaloid biosynthesis\n",
      "Pathway id and name: 00960, Tropane, piperidine and pyridine alkaloid biosynthesis\n",
      "Pathway id and name: 00996, Biosynthesis of various alkaloids\n",
      "Pathway id and name: 00232 , Caffeine metabolism\n",
      "Pathway id and name: 00965 , Betalain biosynthesis\n",
      "Pathway id and name: 00966 , Glucosinolate biosynthesis\n",
      "Pathway id and name: 00402, Benzoxazinoid biosynthesis\n",
      "Pathway id and name: 00311 , Penicillin and cephalosporin biosynthesis\n",
      "Pathway id and name: 00332 , Carbapenem biosynthesis\n",
      "Pathway id and name: 00261 , Monobactam biosynthesis\n",
      "Pathway id and name: 00331 , Clavulanic acid biosynthesis\n",
      "Pathway id and name: 00521 , Streptomycin biosynthesis\n",
      "Pathway id and name: 00524, Neomycin, kanamycin and gentamicin biosynthesis\n",
      "Pathway id and name: 00525 , Acarbose and validamycin biosynthesis\n",
      "Pathway id and name: 00401, Novobiocin biosynthesis\n",
      "Pathway id and name: 00404 , Staurosporine biosynthesis\n",
      "Pathway id and name: 00405 , Phenazine biosynthesis\n",
      "Pathway id and name: 00333 , Prodigiosin biosynthesis\n",
      "Pathway id and name: 00254 , Aflatoxin biosynthesis\n",
      "Pathway id and name: 00998 , Biosynthesis of various antibiotics\n",
      "Pathway id and name: 00999 , Biosynthesis of various plant secondary metabolites\n",
      "Pathway id and name: 00997 , Biosynthesis of various other secondary metabolites\n",
      "b node = 1.11 Xenobiotics biodegradation and metabolism\n",
      "Pathway class number and name: 1.11, Xenobiotics\n",
      "Pathway id and name: 00362 , Benzoate degradation\n",
      "Pathway id and name: 00627 , Aminobenzoate degradation\n",
      "Pathway id and name: 00364, Fluorobenzoate degradation\n",
      "Pathway id and name: 00625, Chloroalkane and chloroalkene degradation\n",
      "Pathway id and name: 00361 , Chlorocyclohexane and chlorobenzene degradation\n",
      "Pathway id and name: 00623 , Toluene degradation\n",
      "Pathway id and name: 00622 , Xylene degradation\n",
      "Pathway id and name: 00633, Nitrotoluene degradation\n",
      "Pathway id and name: 00642, Ethylbenzene degradation\n",
      "Pathway id and name: 00643, Styrene degradation\n",
      "Pathway id and name: 00791, Atrazine degradation\n",
      "Pathway id and name: 00930, Caprolactam degradation\n",
      "Pathway id and name: 00363, Bisphenol degradation\n",
      "Pathway id and name: 00621 , Dioxin degradation\n",
      "Pathway id and name: 00626 , Naphthalene degradation\n",
      "Pathway id and name: 00624 , Polycyclic aromatic hydrocarbon degradation\n",
      "Pathway id and name: 00365, Furfural degradation\n",
      "Pathway id and name: 00984, Steroid degradation\n",
      "Pathway id and name: 00980, Metabolism of xenobiotics by cytochrome P450\n",
      "Pathway id and name: 00982, Drug metabolism - cytochrome P450\n",
      "Pathway id and name: 00983, Drug metabolism - other enzymes\n",
      "Number of pathways: 170\n",
      "Carbohydrate, Glycolysis / Gluconeogenesis,00010 \n",
      "Carbohydrate, Citrate cycle (TCA cycle),00020 \n",
      "Carbohydrate, Pentose phosphate pathway,00030 \n",
      "Carbohydrate, Pentose and glucuronate interconversions,00040 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Carbohydrate, Fructose and mannose metabolism,00051 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Carbohydrate, Galactose metabolism,00052 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Carbohydrate, Ascorbate and aldarate metabolism,00053 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Carbohydrate, Starch and sucrose metabolism,00500 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Carbohydrate, Pyruvate metabolism,00620 \n",
      "Carbohydrate, Glyoxylate and dicarboxylate metabolism,00630 \n",
      "Carbohydrate, Propanoate metabolism,00640 \n",
      "Carbohydrate, Butanoate metabolism,00650 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Carbohydrate, C5-Branched dibasic acid metabolism,00660 \n",
      "Carbohydrate, Inositol phosphate metabolism,00562 \n",
      "Energy, Oxidative phosphorylation,00190 \n",
      "Energy, Photosynthesis,00195 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Energy, Photosynthesis - antenna proteins,00196\n",
      "Energy, Carbon fixation by Calvin cycle,00710 \n",
      "Energy, Other carbon fixation pathways,00720 \n",
      "Energy, Methane metabolism,00680 \n",
      "Energy, Nitrogen metabolism,00910 \n",
      "Energy, Sulfur metabolism,00920 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Lipid, Fatty acid biosynthesis,00061 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Lipid, Fatty acid elongation,00062 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Lipid, Fatty acid degradation,00071 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Lipid, Cutin, suberine and wax biosynthesis,00073\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Lipid, Mycolic acid biosynthesis,00074 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Lipid, Steroid biosynthesis,00100 \n",
      "Lipid, Primary bile acid biosynthesis,00120 \n",
      "Lipid, Secondary bile acid biosynthesis,00121\n",
      "Lipid, Steroid hormone biosynthesis,00140 \n",
      "Lipid, Glycerolipid metabolism,00561 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Lipid, Glycerophospholipid metabolism,00564 \n",
      "Lipid, Ether lipid metabolism,00565 \n",
      "Lipid, Sphingolipid metabolism,00600 \n",
      "Lipid, Arachidonic acid metabolism,00590\n",
      "Lipid, Linoleic acid metabolism,00591\n",
      "Lipid, alpha-Linolenic acid metabolism,00592 \n",
      "Lipid, Biosynthesis of unsaturated fatty acids,01040 \n",
      "Nucleotide, Purine metabolism,00230 \n",
      "Skipping a compound because RDKit can't handle it.\n",
      "Skipping a compound because RDKit can't handle it.\n"
     ]
    }
   ],
   "source": [
    "##################################################################\n",
    "# MAIN\n",
    "##################################################################\n",
    "#compoundDict = {}\n",
    "#moduleDict = {}\n",
    "\n",
    "\n",
    "#\n",
    "# Assemble database of compounds in compoundDict.\n",
    "#\n",
    "pathways = getPathways()\n",
    "print(\"Number of pathways: \" + str(len(pathways)))\n",
    "#for pw in pathways:\n",
    "    \n",
    "#\n",
    "# Convert the list of modules to a CSV.\n",
    "#\n",
    "with open('keggdb2.csv', 'w') as csvfile:\n",
    "    headers = ['ID', 'SMILES', 'MACCS', 'Pathway ID', 'Pathway name', 'Pathway class']\n",
    "    writer = csv.DictWriter(csvfile, fieldnames=headers)\n",
    "    writer.writeheader()\n",
    "    for mod in pathways:\n",
    "        print(mod.className + \", \" + mod.name + \",\" + mod.id)\n",
    "        compounds = mod.fetchCompounds()\n",
    "        for comp in compounds:\n",
    "            #print(comp.fetchMOLFileString())\n",
    "            #print(comp.getSMILES())\n",
    "    #        print(comp.getMACCS())\n",
    "    #        smiles = comp.getSMILES()\n",
    "    #        if (smiles not in compoundDict):\n",
    "    #            compoundDict[smiles] = comp\n",
    "            comp.addPathway(mod)\n",
    "        #for comp in mod.compounds:\n",
    "            if (\"Unhandled\" in comp.getSMILES()):\n",
    "                # Skip this compound, as rdkit can't handle it.\n",
    "                print(\"Skipping a compound because RDKit can't handle it.\")\n",
    "                continue\n",
    "            #if (comp.id in compoundWrittenSet):\n",
    "            #    continue\n",
    "            # Indicate that this compound has already been written.\n",
    "            #compoundWrittenSet.add(comp.id)\n",
    "            \n",
    "            #TODO: Write the row for this compound.\n",
    "            writer.writerow({'ID':comp.id, 'SMILES':comp.getSMILES(), 'MACCS':comp.getMACCS(), 'Pathway ID':mod.id, 'Pathway name':mod.name, 'Pathway class':mod.className })\n",
    "    \n",
    "    \n",
    "            #for modComp in comp.modules:\n",
    "                #TODO: Append to the modules string some representation of\n",
    "                # these modules..\n",
    "            #print(comp.getSMILES() + \",\" + \n",
    "            #print(comp.getMACCS())\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "chem_gcn",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
